dbms/src/Flash/Planner/PhysicalPlanBuilder.h

```diff
@@ -29,18 +31,27 @@ class PhysicalPlanBuilder
         , log(Logger::get("PhysicalPlanBuilder", req_id))
     {}
 
-    void buildSource(const Block & sample_block);
+    void build(const tipb::DAGRequest * dag_request);
 
-    PhysicalPlanPtr getResult() const
-    {
-        RUNTIME_ASSERT(cur_plans.size() == 1, log, "There can only be one plan output, but here are {}", cur_plans.size());
-        return cur_plans.back();
-    }
+    void build(const String & executor_id, const tipb::Executor * executor);
+
+    void buildSource(const BlockInputStreams & source_streams);
```

Comment:
seems `buildSource` is only for the `virtual source` like `project` and `join`? In my understanding, for `tablescan` and `exchange receiver`, there is no source? Is `buildSource` just a tmp function that work with `QueryBlock`?

Reply:
Yes, after `DAGQueryBlock` is removed, `PhysicalSource` is also removed.